lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,625 @@
+        // reference preserved in the IR, such that ExportVerilog can insert a
+        // proper name here at the last moment.
+        auto name = op->getAttrOfType<StringAttr>("name");
+        if (!name) {
+          auto diag = op->emitError("target of memory tap must have a name");
+          diag.attachNote(blackBox.extModule->getLoc())
+              << "used by MemTapAnnotation on port " << portName << " here:";
+          signalPassFailure();
+          continue;
+        }
+
+        // Formulate a hierarchical reference into the memory.
+        // CAVEAT: This just assumes that the memory will map to something that
+        // can be indexed in the final Verilog. If the memory gets turned into
+        // an instance of some sort, we lack the information necessary to go in
+        // and access individual elements of it. This will break horribly since
+        // we generate memory impls out-of-line already, and memories coming
+        // from an external generator are even worse. This needs a special node
+        // in the IR that can properly inject the memory array on emission.
+        wiring.prefices =
+            instanceGraph.getAbsolutePaths(op->getParentOfType<FModuleOp>());
+        wiring.suffix = name.getValue();
+        wiring.suffix += '[';
+        wiring.suffix += llvm::utostr(memPortIdx[portAnno.anno.getDict()]++);
+        wiring.suffix += ']';
+        portWiring.push_back(std::move(wiring));
+        continue;
+      }
+
+      llvm_unreachable("only accepted annos added to portAnnos list");
+    }
+
+#ifndef NDEBUG
+    LLVM_DEBUG(llvm::dbgs() << "- Wire up as follows:\n");
+    for (auto wiring : portWiring) {
+      LLVM_DEBUG(llvm::dbgs() << "- Port " << wiring.portNum << ":\n");
+      for (auto path : wiring.prefices) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "  - " << path << "." << wiring.suffix << "\n");
+      }
+    }
+#endif
+
+    // Now we have an awkward mapping problem. We have multiple data tap module
+    // instances, which reference things in modules that in turn have multiple
+    // instances. This is a side-effect of how Grand Central annotates things on
+    // modules rather than instances. (However in practice these will have a
+    // one-to-one correspondence due to CHIRRTL having fully uniquified
+    // instances.) To solve this issue, create a dedicated implementation for
+    // every data tap instance, and among the possible targets for the data taps
+    // choose the one with the shortest relative path to the data tap instance.
+    OpBuilder builder(blackBox.extModule);
+    unsigned implIdx = 0;
+    for (auto path : paths) {
+      builder.setInsertionPointAfter(blackBox.extModule);
+
+      // Create a new firrtl.module that implements the data tap.
+      SmallString<32> nameBuffer(blackBox.extModule.getName());
+      nameBuffer += "_impl_";
+      nameBuffer += llvm::utostr(implIdx++);
+      auto name = StringAttr::get(&getContext(), nameBuffer);
```

Comment:
I think this would be simpler and just as fast with:

`(Twine(blackBox.extModule.getName()) + "_impl_" + Twine(implIdx++)).str()`

We should really add a StringAttr::get that takes a twine tho